Adds support for the IPFS protocols: (#8805)
curl/curl: docs/IPFS.md

# IPFS

> For an overview about IPFS, visit the [IPFS project site](https://ipfs.tech/).

Reviewer: You will need to extend `.github/scripts/spellcheck.words` with IPFS and IPNS at least, to please the spellchecker.

Author: Thanx for that pointer, will update that one too.

In IPFS there are two protocols. IPFS and IPNS (their workings are explained in detail [here](https://docs.ipfs.tech/concepts/)). The ideal way to access data on the IPFS network is trough those protocols. For example to access the Big Buck Bunny video the ideal way to access it is like: `ipfs://bafybeigagd5nmnn2iys2f3doro7ydrevyr2mzarwidgadawmamiteydbzi`

## IPFS and IPNS protocol handling

> There are various ways to access data from the IPFS network. One such way is through the concept of public "[gateways](https://docs.ipfs.tech/concepts/ipfs-gateway/#overview)". The short version is that entities can offer gateway services. An example here that is hosted by Protocol Labs (who also makes IPFS) is dweb.link and ipfs.io. Both sites expose gateway functionality. Getting a file through ipfs.io looks like this: https://ipfs.io/ipfs/bafybeigagd5nmnn2iys2f3doro7ydrevyr2mzarwidgadawmamiteydbzi

Reviewer: No backticks around the URL

> If you were to be [running your own IPFS node](https://docs.ipfs.tech/how-to/command-line-quick-start/) then you, by default, also have a [local gateway](https://docs.ipfs.tech/concepts/ipfs-gateway/#gateway-providers) running. In it's default configuration the earlier example would then also work in this link: http://127.0.0.1:8080/ipfs/bafybeigagd5nmnn2iys2f3doro7ydrevyr2mzarwidgadawmamiteydbzi

Reviewer: no backticks

Author: ? There is no backtick in the section/line you comment on. The only one i saw (and removed locally) is the one on line 4.

Or do you mean that all "code blocks" that have \``` and end with \``` should all be removed?

Reviewer: I mean you should write the URL within backticks:
`http://127.0.0.1:8080/ipfs/bafybeigagd5nmnn2iys2f3doro7ydrevyr2mzarwidgadawmamiteydbzi`

Author: Right, fixed all of them.

> If you were to be [running your own IPFS node](https://docs.ipfs.tech/how-to/command-line-quick-start/) then you, by default, also have a [local gateway](https://docs.ipfs.tech/concepts/ipfs-gateway/#gateway-providers) running. In it's default configuration the earlier example would then also work in this link: `http://127.0.0.1:8080/ipfs/bafybeigagd5nmnn2iys2f3doro7ydrevyr2mzarwidgadawmamiteydbzi`

Reviewer 2: nit: 
- Kubo is one of many implementations, let's make it more obvious
- linking to specs may be more future-proof

```suggestion
If you were to be [running your own IPFS node (like Kubo)](https://docs.ipfs.tech/how-to/command-line-quick-start/) then you, by default, also have a [local gateway](https://specs.ipfs.tech/http-gateways/) running. In it's default configuration the earlier example would then also work in this link: `http://127.0.0.1:8080/ipfs/bafybeigagd5nmnn2iys2f3doro7ydrevyr2mzarwidgadawmamiteydbzi`
```

## cURL handling of the IPFS protocols

> The IPFS integration in cURL hides this gateway logic for you. So instead of providing a full url to a file on IPFS like this:

Reviewer 2: ```suggestion
The IPFS integration in cURL hides this gateway logic for you. So instead of providing a full URL to a file on IPFS like this:
```

2. Look for the file: `~/.ipfs/gateway`. If it can find that file then it means that you have a local gateway running and that file contains the URL to your local gateway.

If cURL fails you'll be presented with an error message and a link to this page to the option most applicable to solving the issue.

### --ipfs-gateway argument
You can also provide a `--ipfs-gateway` argument to cURL. This overrules any other gateway setting. curl won't fallback to the other options if the provided gateway didn't work.

## Gateway redirects
A gateway could redirect to another place. For example, `dweb.link` redirects [path based](https://docs.ipfs.tech/how-to/address-ipfs-on-web/#path-gateway) requests to [subdomain based](https://docs.ipfs.tech/how-to/address-ipfs-on-web/#subdomain-gateway) ones. So a request to:
```
curl ipfs://bafybeigagd5nmnn2iys2f3doro7ydrevyr2mzarwidgadawmamiteydbzi --ipfs-gateway https://dweb.link
```
Which would be translated to:
```
https://dweb.link/ipfs/bafybeigagd5nmnn2iys2f3doro7ydrevyr2mzarwidgadawmamiteydbzi
```
Will redirect to:
```
https://bafybeigagd5nmnn2iys2f3doro7ydrevyr2mzarwidgadawmamiteydbzi.ipfs.dweb.link
```

> If you thrust this behavior from your gateway of choice then passing the `-L` option will follow the redirect.

Reviewer 3: ```suggestion
If you trust this behavior from your gateway of choice then passing the `-L` option will follow the redirect.
```

## Error messages and hints
Depending on the arguments, cURL could present the user with an error.

### Gateway file and environment variable
cURL tried to look for the file: `~/.ipfs/gateway` but couldn't find it. It also tried to look for the `IPFS_GATEWAY` environment variable but couldn't find that either. This happens when no extra arguments are passed to cURL and letting it try to figure it out [automatically](#Automatic-gateway-detection).

> You could fix this a couple different ways. The easiest is to run a [KUBO node](https://github.com/ipfs/kubo/releases) of version 0.15.0 or higher. Running an up to date version of [IPFS Desktop](https://github.com/ipfs/ipfs-desktop/releases) should fix it too.

Reviewer: This feels like recommendations that will grow old very fast. Can you rephrase this without mentioning any specific product or version number? The advice is to run an IPFS gateway of your own isn't it? It is not 100% clear to me that "KUBO" and "IPFS Desktop" are examples of such.

But also: just because curl doesn't find the file or the environment variable does not mean that there is no gateway running. It is more like forgetting to put the sign up where to find it. Isn't it?

If cURL fails you'll be presented with an error message and a link to this page to the option most applicable to solving the issue.

### --ipfs-gateway argument

> You can also provide a `--ipfs-gateway` argument to cURL. This overrules any other gateway setting. CURL won't fallback to the other options if the provided gateway didn't work.

Reviewer: We call it "curl".

Author: That fails your checks. A line must start with a capital letter. I'm fine calling it Curl, curl, cURL or even a mixture thereof. As long as it doesn't fail compilation.

Reviewer: No, because `curl` and `libcurl` are accepted as-is after periods in markdown files.

Reviewer: If not, the test is buggy.

Author: Ohh sweet! Will change then :)
